File: src/Fable.Electron.Remoting/Renderer.js

import { Record } from "../../tests/Fable.Electron.Remoting.Tests/fable_modules/fable-library-js.5.0.0-alpha.14/Types.js";
import { fullName, makeRecord, name, record_type, lambda_type, string_type } from "../../tests/Fable.Electron.Remoting.Tests/fable_modules/fable-library-js.5.0.0-alpha.14/Reflection.js";
import { toFail, toText } from "../../tests/Fable.Electron.Remoting.Tests/fable_modules/fable-library-js.5.0.0-alpha.14/String.js";
import { createTypeInfo } from "../../tests/Fable.Electron.Remoting.Tests/fable_modules/Fable.SimpleJson.3.24.0/TypeInfo.Converter.fs.js";
import { map, delay, toArray } from "../../tests/Fable.Electron.Remoting.Tests/fable_modules/fable-library-js.5.0.0-alpha.14/Seq.js";
import { item } from "../../tests/Fable.Electron.Remoting.Tests/fable_modules/fable-library-js.5.0.0-alpha.14/Array.js";

export class RemotingConfig extends Record {
    constructor(ApiNameBase, ApiNameMap, ChannelNameMap) {
        super();
        this.ApiNameBase = ApiNameBase;
        this.ApiNameMap = ApiNameMap;
        this.ChannelNameMap = ChannelNameMap;
    }
}

export function RemotingConfig_$reflection() {
    return record_type("Fable.Electron.Remoting.Renderer.RemotingConfig", [], RemotingConfig, () => [["ApiNameBase", string_type], ["ApiNameMap", lambda_type(string_type, lambda_type(string_type, string_type))], ["ChannelNameMap", lambda_type(string_type, lambda_type(string_type, string_type))]]);
}

export const RemotingModule_init = new RemotingConfig("FABLE_REMOTING", (baseName, typeName) => toText(`${baseName}_${typeName}`), (typeName_1, fieldName) => toText(`${typeName_1}:${fieldName}`));

export function RemotingModule_withApiNameBase(apiName, config) {
    return new RemotingConfig(apiName, config.ApiNameMap, config.ChannelNameMap);
}

export function RemotingModule_withApiNameMap(func, config) {
    return new RemotingConfig(config.ApiNameBase, func, config.ChannelNameMap);
}

export function RemotingModule_withChannelNameMap(func, config) {
    return new RemotingConfig(config.ApiNameBase, config.ApiNameMap, func);
}

export function Remoting_buildProxySender_69894920(config, resolvedType) {
    const schemaType = createTypeInfo(resolvedType);
    if (schemaType.tag === 39) {
        const getFields = schemaType.fields[0];
        const patternInput = getFields();
        const recordType = patternInput[1];
        const fields = patternInput[0];
        const bridgeName = config.ApiNameMap(config.ApiNameBase, name(resolvedType));
        const recordFields = toArray(delay(() => map((field) => ((window[bridgeName])[field.FieldName]), fields)));
        const proxy = makeRecord(recordType, recordFields);
        return proxy;
    }
    else {
        return toFail(`Cannot build proxy. Expected type ${fullName(resolvedType)} to be a valid protocol definition which is a record of functions`);
    }
}

export function Remoting_buildProxyReceiver_69F94225(impl, config, resolvedType) {
    const schemaType = createTypeInfo(resolvedType);
    const bridgeName = config.ApiNameMap(config.ApiNameBase, name(resolvedType));
    if (schemaType.tag === 39) {
        const getFields = schemaType.fields[0];
        const patternInput = getFields();
        const recordType = patternInput[1];
        const fields = patternInput[0];
        for (let idx = 0; idx <= (fields.length - 1); idx++) {
            const field = item(idx, fields);
            const callSite = (window[bridgeName])[field.FieldName];
            const fieldTarget = impl[field.FieldName];
            const func = callSite((...args) => { return fieldTarget(...args) });
        }
    }
    else {
        toFail(`Cannot build proxy. Expected type ${fullName(resolvedType)} to be a valid protocol definition which is a record of functions`);
    }
}

File: src/Fable.Electron.Remoting/Renderer.fs
module Fable.Electron.Remoting.Renderer

open System
open System.ComponentModel
open Browser
open FSharp.Core
open Fable.Core
open Fable.Core.DynamicExtensions
open FSharp.Reflection
open Fable.Core.JsInterop
open Fable.SimpleJson
//%REMOTING_TYPE%START%
/// <summary>
/// Config for a proxy IPC router.
/// </summary>
type RemotingConfig = {
    /// <summary>
    /// An argument to the <c>ApiNameMap</c> that creates the name of the property on the
    /// <c>window</c> object that the proxy/api/IPC is exposed through.
    /// </summary>
    ApiNameBase: string
    /// <summary>
    /// <c>ApiNameMap</c> takes the <c>ApiNameBase</c> and <c>Type</c> name of the
    /// implementation to create the name of the property on the
    /// <c>window</c> object that the proxy/api/IPC is exposed through.
    /// </summary>
    ApiNameMap: string -> string -> string
    /// <summary>
    /// No effect on Renderer process.
    /// </summary>
    ChannelNameMap: string -> string -> string
}
//%REMOTING_TYPE%END%
[<Erase>]
module Remoting =
    let init = {
        ApiNameBase = "FABLE_REMOTING"
        ApiNameMap = fun baseName typeName -> sprintf $"%s{baseName}_{typeName}"
        ChannelNameMap = fun typeName fieldName -> sprintf $"%s{typeName}:%s{fieldName}"
    }

    let withApiNameBase apiName config = { config with ApiNameBase = apiName }
    let withApiNameMap func config = { config with ApiNameMap = func }
    let withChannelNameMap func config = { config with ChannelNameMap = func }

[<Erase>]
type Remoting =
    [<EditorBrowsable(EditorBrowsableState.Never)>]
    static member buildProxySender(config: RemotingConfig, resolvedType: Type) =
        let schemaType = createTypeInfo resolvedType

        match schemaType with
        | TypeInfo.Record getFields ->
            let fields, recordType = getFields ()

            let bridgeName = config.ApiNameMap config.ApiNameBase resolvedType.Name

            let recordFields = [|
                for field in fields do
                    match window.Item(bridgeName) with
                    | null ->
                        failwith $"No API found on window for bridge name {bridgeName}. This might be because a related bridge in preload is missing."
                    | bridge ->
                        match bridge.Item(field.FieldName) with
                        | null ->
                            failwith $"No API found on bridge {bridgeName} for field {field.FieldName}"
                        | apiFn ->
                            apiFn
            |]

            let proxy = FSharpValue.MakeRecord(recordType, recordFields)
            unbox proxy
        | _ ->
            failwithf
                $"Cannot build proxy. Expected type %s{resolvedType.FullName} to be \
                a valid protocol definition which is a record of functions"

    [<EditorBrowsable(EditorBrowsableState.Never)>]
    static member buildProxyReceiver(impl, config: RemotingConfig, resolvedType: Type) : unit -> unit =
        let schemaType = createTypeInfo resolvedType
        let bridgeName = config.ApiNameMap config.ApiNameBase resolvedType.Name

        match schemaType with
        | TypeInfo.Record getFields ->
            let fields, recordType = getFields ()
            let disposers = ResizeArray<unit -> unit>()

            for field in fields do
                let callSite = window.Item(bridgeName).Item(field.FieldName)
                let fieldTarget = impl.Item(field.FieldName)

                let disposer: unit -> unit =
                    emitJsExpr (callSite, fieldTarget, impl) "$0((...args) => { return $1(...args) })"

                disposers.Add disposer

            let mutable disposed = false

            fun () ->
                if not disposed then
                    disposed <- true

                    for dispose in disposers do
                        dispose ()
        | _ ->
            failwithf
                $"Cannot build proxy. Expected type %s{resolvedType.FullName} to be \
                a valid protocol definition which is a record of functions"

    /// <summary>
    /// Builds the client for the <c>Main &lt;-> Renderer</c> two way IPC communication.
    /// </summary>
    /// <param name="config"></param>
    static member inline buildClient<'t>(config: RemotingConfig) : 't =
        Remoting.buildProxySender (config, typeof<'t>)

    /// <summary>
    /// Builds the receiver for the <c>Main -> Renderer</c> IPC communication.
    /// </summary>
    /// <param name="impl">The implemented record of functions that respond to messages.</param>
    /// <param name="config"></param>
    static member inline buildHandlerDisposable<'T> (impl: 'T) (config: RemotingConfig) : unit -> unit =
        Remoting.buildProxyReceiver (impl, config, typeof<'T>)

    /// <summary>
    /// Builds the receiver for the <c>Main -> Renderer</c> IPC communication.
    /// </summary>
    /// <param name="impl">The implemented record of functions that respond to messages.</param>
    /// <param name="config"></param>
    static member inline buildHandler<'T> (impl: 'T) (config: RemotingConfig) =
        Remoting.buildProxyReceiver (impl, config, typeof<'T>) |> ignore